web/src/message_events.js

```diff
@@ -35,6 +37,129 @@ import * as unread_ops from "./unread_ops";
+    // can be applied locally.
+    const supported_term_types = [
+        "has-image",
+        "has-link",
+        "has-reaction",
+        "has-attachment",
+        "is-starred",
+        "is-unread",
+        "is-mentioned",
+        "is-alerted",
+        // TODO: Implement support for these terms.
+        // "is-followed",
+    ];
+
+    if (message_ids.length === 0 || !supported_term_types.includes(term_type)) {
+        return;
+    }
+
+    for (const msg_list of message_lists.all_rendered_message_lists()) {
+        const filter = msg_list.data.filter;
+        const filter_term_types = filter.sorted_term_types();
+        if (
+            // Check if current filter relies on the changed message property.
+            !filter_term_types.includes(term_type) &&
+            !filter_term_types.includes(`not-${term_type}`)
+        ) {
+            continue;
+        }
+
+        if (!filter.can_apply_locally()) {
+            // We can't apply the update locally, so we need to rerender the view.
+            const current_selected_id = msg_list.selected_id();
+            message_view.show(filter.terms(), {
+                then_select_id: current_selected_id,
+                trigger: "message property update",
+                force_rerender: true,
+            });
+            continue;
+        }
+
+        // We need the message objects to determine if they match the filter.
+        const messages_to_fetch = [];
+        const messages = [];
+        for (const message_id of message_ids) {
+            const message = message_store.get(message_id);
+            if (message !== undefined) {
+                messages.push(message);
+            } else {
+                const first_id = msg_list.first().id;
+                const last_id = msg_list.last().id;
+                const has_found_newest = msg_list.data.fetch_status.has_found_newest();
+                const has_found_oldest = msg_list.data.fetch_status.has_found_oldest();
+
+                if (message_id > first_id && message_id < last_id) {
+                    // Need to insert message middle of the list.
+                    messages_to_fetch.push(message_id);
+                } else if (message_id < first_id && has_found_oldest) {
+                    // Need to insert message at the start of list.
+                    messages_to_fetch.push(message_id);
+                } else if (message_id > last_id && has_found_newest) {
+                    // Need to insert message at the end of list.
+                    messages_to_fetch.push(message_id);
+                }
+            }
+        }
+
+        // In most cases, we will only have one message to fetch which
+        // can happen without rerendering the view.
+        // In case of multiple messages, we just rerender the view
+        // since it is likely to use similar amount of resources to
+        // fetching the messages and rerendering the view.
```

Comment:
I assume part of the issue here is we don't have a good API endpoint for fetching several messages by ID, right?

Reply:
Yeah.